File: backend/api/_data_loader.py
"""
Shared data loading module that works with Vercel Blob
"""
import csv
import os
from io import StringIO

def load_csv_from_blob(filename):
    """
    Load CSV from Vercel Blob storage
    In production, this will fetch from Vercel Blob
    In development, falls back to local files
    """
    blob_token = os.environ.get('BLOB_READ_WRITE_TOKEN')

    if blob_token:
        try:
            import requests

            # Vercel Blob API endpoint
            api_url = "https://blob.vercel-storage.com"
            headers = {"Authorization": f"Bearer {blob_token}"}

            # First, list blobs to find the one we want
            list_response = requests.get(
                api_url,
                headers=headers,
                params={"prefix": f"data/{filename}"}
            )

            if list_response.status_code == 200:
                data = list_response.json()
                blobs = data.get('blobs', [])

                if blobs and len(blobs) > 0:
                    # Get the download URL from the blob
                    blob_url = blobs[0].get('downloadUrl') or blobs[0].get('url')

                    if blob_url:
                        # Fetch the CSV content
                        download_response = requests.get(blob_url)
                        if download_response.status_code == 200:
                            content = download_response.text
                            reader = csv.DictReader(StringIO(content))
                            return list(reader)

        except Exception as e:
            # Log error for debugging
            print(f"Blob loading error for {filename}: {e}")
            import traceback
            traceback.print_exc()

    # Fallback to local files for development
    try:
        data_dir = os.path.join(os.path.dirname(__file__), '../../data')
        filepath = os.path.join(data_dir, filename)

        with open(filepath, 'r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            return list(reader)
    except Exception as e:
        print(f"Local file loading error for {filename}: {e}")
        return []

def parse_fund_simple(rows, fund_name):
    """Parse fund data from CSV rows"""
    if not rows:
        return {'fund_name': fund_name, 'holdings': [], 'total_stocks': 0}

    keys = rows[0].keys()

    if 'Company Name' in keys:
        name_col = 'Company Name'
        weight_col = '% Portfolio Weight'
    else:
        name_col = 'Holding Name'
        weight_col = '% Portfolio'

    holdings = []
    for row in rows:
        try:
            name = row.get(name_col, '').strip()
            weight_str = row.get(weight_col, '0')

            if not name or name == '':
                continue

            weight_str = str(weight_str).replace('%', '').strip()
            if weight_str == '':
                continue

            weight = float(weight_str)

            holdings.append({
                'name': name,
                'weight': round(weight, 2)
            })
        except:
            continue

    # Sort alphabetically by name
    holdings.sort(key=lambda x: x['name'])

    return {
        'fund_name': fund_name,
        'holdings': holdings,
        'total_stocks': len(holdings)
    }

def get_available_funds():
    """Get list of available funds"""
    blob_token = os.environ.get('BLOB_READ_WRITE_TOKEN')

    if blob_token:
        try:
            import requests

            # Vercel Blob API endpoint
            api_url = "https://blob.vercel-storage.com"
            headers = {"Authorization": f"Bearer {blob_token}"}

            # List all blobs with 'data/' prefix
            list_response = requests.get(
                api_url,
                headers=headers,
                params={"prefix": "data/", "limit": "1000"}
            )

            if list_response.status_code == 200:
                data = list_response.json()
                blobs = data.get('blobs', [])

                funds = []
                for blob in blobs:
                    pathname = blob.get('pathname', '')
                    # Extract filename from pathname (e.g., 'data/fund.csv' -> 'fund.csv')
                    if pathname.startswith('data/') and pathname.endswith('.csv'):
                        filename = pathname.replace('data/', '', 1)
                        fund_id = filename.replace('.csv', '')
                        funds.append({
                            'id': fund_id,
                            'name': fund_id.replace('_', ' ').title()
                        })

                return funds
            else:
                print(f"Blob API error: {list_response.status_code} - {list_response.text}")

        except Exception as e:
            print(f"Error listing funds from blob: {e}")
            import traceback
            traceback.print_exc()

    # Fallback to local files for development
    try:
        data_dir = os.path.join(os.path.dirname(__file__), '../../data')
        files = [f for f in os.listdir(data_dir) if f.endswith('.csv')]
        return [
            {
                'id': f.replace('.csv', ''),
                'name': f.replace('.csv', '').replace('_', ' ').title()
            }
            for f in files
        ]
    except Exception as e:
        print(f"Error listing local files: {e}")
        return []

File: lib/dataLoader.ts
import { list } from '@vercel/blob';
import fs from 'fs';
import path from 'path';
import { parse } from 'csv-parse/sync';

export interface Holding {
  name: string;
  weight: number;
}

export interface FundData {
  fund_name: string;
  holdings: Holding[];
  total_stocks: number;
}

export interface Fund {
  id: string;
  name: string;
}

async function loadCsvFromBlob(filename: string): Promise<any[]> {
  const token = process.env.BLOB_READ_WRITE_TOKEN;

  if (token) {
    try {
      // List blobs to find the file
      const { blobs } = await list({
        prefix: `data/${filename}`,
        token,
      });

      if (blobs && blobs.length > 0) {
        const blobUrl = blobs[0].url;

        // Fetch CSV content
        const response = await fetch(blobUrl);
        if (response.ok) {
          const content = await response.text();
          const records = parse(content, {
            columns: true,
            skip_empty_lines: true,
            relax_column_count: true,
            trim: true,
          });
          return records;
        }
      }
    } catch (error) {
      console.error(`Blob loading error for ${filename}:`, error);
    }
  }

  // Fallback to local files for development
  try {
    const dataDir = path.join(process.cwd(), 'data');
    const filepath = path.join(dataDir, filename);
    const content = fs.readFileSync(filepath, 'utf-8');
    const records = parse(content, {
      columns: true,
      skip_empty_lines: true,
      relax_column_count: true,
      trim: true,
    });
    return records;
  } catch (error) {
    console.error(`Local file loading error for ${filename}:`, error);
    return [];
  }
}

function parseFund(rows: any[], fundName: string): FundData {
  if (!rows || rows.length === 0) {
    return { fund_name: fundName, holdings: [], total_stocks: 0 };
  }

  const keys = Object.keys(rows[0]);
  const nameCol = keys.includes('Company Name') ? 'Company Name' : 'Holding Name';
  const weightCol = keys.includes('% Portfolio Weight') ? '% Portfolio Weight' : '% Portfolio';

  const holdings: Holding[] = [];

  for (const row of rows) {
    try {
      const name = row[nameCol]?.trim();
      let weightStr = row[weightCol];

      if (!name) continue;

      weightStr = String(weightStr).replace('%', '').trim();
      if (!weightStr) continue;

      const weight = parseFloat(weightStr);
      holdings.push({
        name,
        weight: Math.round(weight * 100) / 100,
      });
    } catch (error) {
      continue;
    }
  }

  // Sort alphabetically by name
  holdings.sort((a, b) => a.name.localeCompare(b.name));

  return {
    fund_name: fundName,
    holdings,
    total_stocks: holdings.length,
  };
}

export async function getAvailableFunds(): Promise<Fund[]> {
  const token = process.env.BLOB_READ_WRITE_TOKEN;

  if (token) {
    try {
      // List all blobs with 'data/' prefix
      const { blobs } = await list({
        prefix: 'data/',
        token,
      });

      const funds: Fund[] = [];
      for (const blob of blobs) {
        const pathname = blob.pathname;
        if (pathname.startsWith('data/') && pathname.endsWith('.csv')) {
          const filename = pathname.replace('data/', '');
          const fundId = filename.replace('.csv', '');
          funds.push({
            id: fundId,
            name: fundId.replace(/_/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase()),
          });
        }
      }

      return funds;
    } catch (error) {
      console.error('Error listing funds from blob:', error);
    }
  }

  // Fallback to local files for development
  try {
    const dataDir = path.join(process.cwd(), 'data');
    const files = fs.readdirSync(dataDir).filter((f) => f.endsWith('.csv'));
    return files.map((f) => ({
      id: f.replace('.csv', ''),
      name: f.replace('.csv', '').replace(/_/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase()),
    }));
  } catch (error) {
    console.error('Error listing local files:', error);
    return [];
  }
}

export async function loadFund(fundId: string): Promise<FundData> {
  const filename = `${fundId}.csv`;
  const rows = await loadCsvFromBlob(filename);
  return parseFund(rows, fundId);
}
